README
Test::WWW::Mechanize::Mojo - Test::WWW::Mechanize for Mojo

Test::WWW::Mechanize::Mojo is a module to test Mojo applications using
Test::WWW::Mechanize and WWW::Mechanize.

It is based on Test::WWW::Mechanize::Catalyst .

INSTALLATION

To install this module, run the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install


SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the perldoc command.

    perldoc Test::WWW::Mechanize::Mojo

You can also look for information at:

    MetaCPAN
        http://metacpan.org/release/Test-WWW-Mechanize-Mojo

COPYRIGHT AND LICENCE

Copyright (C) 2005-8, Leon Brocard

This module is free software; you can redistribute it or modify it
under the same terms as Perl itself.



Test-WWW-Mechanize-Mojo/CHANGES
Revision history for Perl module Test::WWW::Mechanize::Mojo :

    - Add separate "LICENSE" file (CPANTS).
    - Specify minimum version of perl (CPANTS).

v0.0.14     Fri 29 Nov 09:36:57 IST 2013
    - Remove some old PODded-out code.
        - Caused a false positive at:
        - https://rt.cpan.org/Public/Bug/Display.html?id=91042

v0.0.13     Sat 25 May 23:27:16 IDT 2013
    - Fix the tests for Mojolicious 4.07.

v0.0.12     Tue 21 May 21:10:56 IDT 2013
    - Add t/style-trailing-space.t and get rid of trailing space.

v0.0.11     Wed  1 Aug 20:55:21 IDT 2012
    - Got the POD coverage ( t/pod-coverage.t ) to pass again.

v0.0.10     Sun May 13 23:16:58 IDT 2012
    - Update the module for recent versions of Mojolicious where
    $t->max_redirects() was removed (after being deprecated).
        - Thanks to tempire on #mojo on irc.perl.org for shedding some
        light on this issue.
    - This release coincides with the birthday of Sebastian Riedel
    ( https://metacpan.org/author/SRI ), the creator and maintainer
    of Mojolicious (and Catalyst previously). Happy birthday, Sebastian!

v0.0.9      Thu May 19 23:14:29 IDT 2011
    - Got rid of warnings by switching from $t->client to $t->ua.
        - Thanks to Mirko Westermeier

v0.0.8      Wed Jan 26 14:56:01 IST 2011
    - Convert the mentioning of CATALYST_SERVER to MOJO_SERVER.
        - thanks to Insurgent Software.

v0.0.7      Mon Jan 17 15:55:00 IST 2011
    - Fixed mojjy.pl and the lib/Test/WWW/Mechanize/Mojo.pm code in recent
    Mojo versions.
        - app->start instead of shagadelic.
        - some tweaks to the ->$method->res.

v0.0.6      Fri Sep  3 20:43:26 IDT 2010
    - Fixed mojjy.pl to avoid double encoding in recent Mojo versions.

v0.0.5      Mon Jun 14 21:54:45 IDT 2010
    - Removed the call to Test::Mojo->redirect() which is no longer
    implemented in recent Mojos.

v0.0.4      Wed Mar 17 16:01:20 IST 2010
    - Now evaluating the GET parameters in requests by using URI.pm's
    ->path_query() instead of ->path().

v0.0.3      Thu Mar  4 00:35:10 IST 2010
    - Fix submit_form_ok on the contemporary Mojo::Client . (with a test).

v0.0.2      Tue Mar  2 23:27:43 IST 2010
    - Add a missing empty line before an "=end". Without it, the POD
    was not displayed properly by perldoc and by http://search.cpan.org/ .
    - Add a t/pod-coverage.t test.

v0.0.1      Tue Mar  2 20:59:50 IST 2010
    - Ported to Mojo from Test::WWW::Mechanize::Catalyst
    - Convert from Module::Install to Module::Build .
    - All rights disclaimed.

Revision history for Perl module Test::WWW::Mechanize::Catalyst:

0.51 Mon Mar 16 10:00 GMT 2009
     - Doc updates from thejester
     - User agent fixes from ANDREMAR
     - Fix bug where redirect was followed on a 500 response
     - All remote requests (i.e. CATALYST_SERVER env var) now use our own
       mechanize object, rather than an unconfigurable one from Catalyst:Test

0.50 Tue Feb 17 09:12 GMT 2009
     - Remove warning in HTTP::Cookies
     - Call BUILDALL

0.50_2 Thur Feb 12 09:47 GMT 2009
     - Make t/multi_content_type.t handle case when server cant be started,
       which is almost always due to port in use.

0.50_1 Thur Feb 5 09:02 GMT 2009
     - App classname no longer has to be passed to import:
        $m = T::W::M::C->new(catalyst_app => 'Catty')
       now works.
     - Can now use TWMC two test two different apps in the same perl
       interpreter due to the above change
     - Removed Test::WWW::Mechanize::Catalyst::Aux package as it isn't needed
       any more
     - Add 'host' accessor for white-label testing
     - Moosification
     - Can now test against remote CATALYST_SERVER without having to load the
       app class

0.45 Mon Nov 24 20:39:19 GMT 2008
     - be forwards-compatible with Catalyst 5.80's virtual
       domain testing (thanks Jason Gottshall)

0.44 Mon Oct 27 13:48:22 GMT 2008
     - fix longstanding bug with recent LWP, requiring
       WWW::Mechanize 1.50 (thanks to petdance, mst, dakkar)
     - add machine- and human-readable license, add abstract

0.43 Mon Aug 18 15:42:03 BST 2008
     - add missing prereqs to Catalyst::Plugin::Session::State::Cookie
       and Catalyst::Plugin::Session::Store::Dummy (thanks kd)

0.42 Tue Apr 29 20:25:06 BST 2008
     - stop multi_content_type.t killing smoke testing
       (patch by Andreas König)
     - fix a case where HTTP::Cookies dies when trying to
       extract_cookies (patch by Andreas Marienborg)
     - add Test::Exception as a prerequisite

0.41 Mon Sep 17 20:28:59 BST 2007
     - fix to cope with gzipped content and the test from the
       rt.cpan queue about multiple content types
       (patch by Ash Berlin)

0.40 Tue Aug 21 20:51:13 BST 2007
     - external requests (as per last release) are now only allowed
       if you set allow_external (sorry about that)

0.39 Sat Aug  4 08:01:38 BST 2007
     - external requests are now allowed (patch by Edmund von der Burg)
     - remove Build.PL

0.38 Sat Jun 30 14:07:24 BST 2007
     - document and test that you can use URLs without schema
       or hostname
     - add debug screen error to test diagnostics (patch by
       Jonathan Swartz)
     - add basic authentication support (patch by Gareth Kirwan)
     - add test for charset=utf-8 (patch by Chris Dolan)
     - added CATALYST_SERVER mention in the documentation
       (patch by Kieren Diment)

0.37 Tue Jun  6 08:54:07 BST 2006
     - patch to follow LWP's $m->requests_redirectable() and small
       docpatch (thanks to Daniel McBrearty)
     - mention Catalyst::Test (thanks to guest)

0.36 Mon Apr 17 11:27:17 BST 2006
     - perltidy
     - Catalyst debug screens are now failures (thanks to Kieren Diment)

0.35 Tue Jan 22 17:06:00 GMT 2006
     - handle redirects (patch by Mark Fowler)

0.33 Tue Jun  7 17:38:45 BST 2005
     - we need at least version 1.04 of Test::WWW::Mechanize
       (spotted by Jesse Vincent, patch by Shlomi Fish)

0.32 Tue May  3 16:14:40 BST 2005

Test-WWW-Mechanize-Mojo/t/lib/Catty.pm
package Catty;

use strict;
use warnings;

#use Catalyst;
use Catalyst;
use Cwd;
use MIME::Base64;
use Encode qw//;

our $VERSION = '0.01';

Catty->config(
    name => 'Catty',
    root => cwd . '/t/root',
);
Catty->setup();
Catty->log->levels("fatal");

sub default : Private {
    my ( $self, $context ) = @_;
    my $html = html( "Root", "This is the root page" );
    $context->response->content_type("text/html");
    $context->response->output($html);
}

sub hello : Global {
    my ( $self, $context ) = @_;
    my $str = Encode::encode('utf-8', "\x{263A}"); # ☺
    my $html = html( "Hello", "Hi there! $str" );
    $context->response->content_type("text/html; charset=utf-8");
    $context->response->output($html);
}

# absolute redirect
sub hi : Global {
    my ( $self, $context ) = @_;
    my $where = $context->uri_for('hello');
    $context->response->redirect($where);
    return;
}

# partial (relative) redirect
sub greetings : Global {
    my ( $self, $context ) = @_;
    $context->response->redirect("hello");
    return;
}

# redirect to a redirect
sub bonjour : Global {
    my ( $self, $context ) = @_;
    my $where = $context->uri_for('hi');
    $context->response->redirect($where);
    return;
}

sub check_auth_basic : Global {
    my ( $self, $context ) = @_;

    my $auth = $context->req->headers->authorization;
    ($auth) = $auth =~ /Basic\s(.*)/i;
    $auth = decode_base64($auth);

    if ( $auth eq "user:pass" ) {
        my $html = html( "Auth", "This is the auth page" );
        $context->response->content_type("text/html");
        $context->response->output($html);
        return $context;
    } else {
        my $html = html( "Auth", "Auth Failed!" );
        $context->response->content_type("text/html");
        $context->response->output($html);
        $context->response->status("401");
        return $context;
    }
}

sub redirect_with_500 : Global {
    my ( $self, $c ) = @_;
    $DB::single = 1;
    $c->res->redirect( $c->uri_for("/bonjour"));
    die "erk!";
}

sub die : Global {
    my ( $self, $context ) = @_;
    my $html = html( "Die", "This is the die page" );
    $context->response->content_type("text/html");
    $context->response->output($html);
    die "erk!";
}

sub name : Global {
    my ($self, $c) = @_;

    my $html = html( $c->config->{name}, "This is the die page" );
    $c->response->content_type("text/html");
    $c->response->output($html);
}

sub host : Global {
    my ($self, $c) = @_;

    my $host = $c->req->header('Host') || "<undef>";
    my $html = html( $c->config->{name}, "Host: $host" );
    $c->response->content_type("text/html");
    $c->response->output($html);
}

sub html {
    my ( $title, $body ) = @_;
    return qq{
<html>
<head><title>$title</title></head>
<body>
$body
<a href="/hello/">Hello</a>.
</body></html>
};
}

sub gzipped : Global {
    my ( $self, $c ) = @_;

  # If done properly this test should check the accept-encoding header, but we
  # control both ends, so just always gzip the response.
    require Compress::Zlib;

    my $html = html( "Hello", "Hi there! ☺" );
    $c->response->content_type("text/html; charset=utf-8");
    $c->response->output( Compress::Zlib::memGzip($html) );
    $c->response->content_encoding('gzip');
    $c->response->headers->push_header( 'Vary', 'Accept-Encoding' );
}

sub user_agent : Global {
    my ( $self, $c ) = @_;
    
    my $html = html($c->req->user_agent, $c->req->user_agent);
    $c->response->content_type("text/html; charset=utf-8");
    $c->response->output( $html );
    
}

1;



Test-WWW-Mechanize-Mojo/t/lib/mojjy.pl
#!/usr/bin/env perl

use Mojolicious::Lite;

use MIME::Base64;
use Encode qw//;
use Cwd;

use utf8;

sub html {
    my ( $title, $body ) = @_;
    return qq{
<html>
<head><title>$title</title></head>
<body>
$body
<a href="/hello/">Hello</a>.
</body></html>
};
}

get '/check_auth_basic/' => sub {
    my $self = shift;

    my $auth = $self->req->headers->header("Authorization");
    ($auth) = $auth =~ /Basic\s(.*)/i;
    $auth = decode_base64($auth);

    if ( $auth eq "user:pass" )
    {
        my $html = html( "Auth", "This is the auth page" );
        $self->render_text($html);
        return;
    }
    else
    {
        my $html = html( "Auth", "Auth Failed!" );
        $self->render_text($html, status => "401",);
        return;
    }
};


get "/hi" => sub {
    my $self = shift;

    $self->redirect_to('/hello');

    return;
};

get "/greetings" => sub {
    my $self = shift;

    # This relative URL is something that Catalyst eats and appears
    # in Catty.pm , but Mojo won't accept.
    # -- Shlomi Fish
    # $self->redirect_to('hello');
    $self->redirect_to('/hello');

    return;
};

get "/bonjour" => sub {
    my $self = shift;

    $self->redirect_to('/hi');

    return;
};


get '/hello' => sub {
    my $self = shift;

    my $html = html( "Hello", "Hi there! ☺" ); # ☺
    $self->res->headers->content_type("text/html; charset=utf-8");
    $self->render_text($html);

    return;
};

get '/redirect_with_500' => sub {
    my $self = shift;

    $self->redirect_to('/bonjour');

    die "erk!";
};

get "/die/" => sub {
    my $self = shift;

    my $html = html( "Die", "This is the die page" );
    $self->render_text($html);
    die "erk!";
};

sub _gzipped {
    my $self = shift;

    # If done properly this test should check the accept-encoding header, but we
    # control both ends, so just always gzip the response.
    require Compress::Zlib;

    my $html = html( "Hello", "Hi there! ☺" );

    $self->res->headers->content_type("text/html; charset=utf-8");
    $self->render_text( Compress::Zlib::memGzip($html) );
    $self->res->headers->content_transfer_encoding('gzip');
    $self->res->headers->add( 'Vary', 'Accept-Encoding' );

    return;
}

get "/gzipped/" => \&_gzipped;


get "/user_agent" => sub {
    my $self = shift;

    my $agent = $self->req->headers->user_agent();
    my $html = html($agent, $agent);

    $self->render_text($html);
    $self->res->headers->content_type("text/html; charset=utf-8");

    return;
};


get "/host" => sub {
    my $self = shift;

    my $host = $self->req->headers->header('Host') || "<undef>";
    my $html = html( "Foo", "Host: $host" );
    $self->render_text($html);

    return;
};

post "/form-submit" => sub {
    my $self = shift;

    my $html = html( "Foo", "Your email is " . $self->param("email"));

    $self->render_text($html);

    return;
};

get "/form" => sub {
    my $self = shift;

    $self->render_text(<<'EOF');
<html>
<head><title>Form test</title></head>
<body>
<form id="register" action="/form-submit" method="post">
<table>

<tr>
<td>Email:</td>
<td><input name="email" /></td>
</tr>

<tr>
<td colspan="2">
<input type="submit" value="Submit" />
</td>
</tr>

</table>
</form>

</body></html>

EOF
    return;
};

get '/with-params' => sub {
    my $self = shift;

    $self->render_text(sprintf("[%s]{%s}", $self->param('one'), $self->param('two')));
};

get '/:groovy' => sub {
    my $self = shift;
    $self->render_text($self->param('groovy'), layout => 'funky');
};

get '/' => sub {
    my $self = shift;

    $self->render_text(html("Root", "This is the root page"));

    return;
};

app->start;

=head1 TODO

* Add a status (Not logged-in / Logged in as something) ruler to the top.

=cut

__DATA__

@@ layouts/funky.html.ep
<!doctype html><html>
    <head><title>Foo Bar</title></head>
    <body><%== content %></body>
</html>


Test-WWW-Mechanize-Mojo/Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.4007
require 5.012;
use ExtUtils::MakeMaker;
WriteMakefile
(
  'NAME' => 'Test::WWW::Mechanize::Mojo',
  'VERSION_FROM' => 'lib/Test/WWW/Mechanize/Mojo.pm',
  'PREREQ_PM' => {
                   'Carp' => 0,
                   'Encode' => 0,
                   'HTML::Entities' => 0,
                   'LWP' => '5.816',
                   'Mojolicious' => '4.75',
                   'Mojolicious::Lite' => 0,
                   'Test::WWW::Mechanize' => '1.14',
                   'WWW::Mechanize' => '1.50',
                   'base' => 0,
                   'strict' => 0,
                   'utf8' => 0,
                   'warnings' => 0
                 },
  'INSTALLDIRS' => 'site',
  'EXE_FILES' => [],
  'PL_FILES' => {}
)
;


Test-WWW-Mechanize-Mojo/Build.PL
use lib "./buildlib";

use Test::Run::Builder;

use strict;
use warnings;

my $build = Test::Run::Builder->new(
    module_name => "Test::WWW::Mechanize::Mojo",
    configure_requires =>
    {
        'Module::Build' => '0.36',
    },
    test_requires =>
    {
        'Test::Exception'=> '0',
        'Test::More' => '0',
        'Test::Mojo' => 0,
    },
    requires =>
    {
        'base' => 0,
        'Carp' => 0,
        'Encode' => 0,
        'HTML::Entities'                           => 0,
        'LWP'                                      => '5.816',
        'Mojolicious::Lite' => 0,
        'Mojolicious' => '4.07',
        'perl' => '5.012',
        'strict' => 0,
        'Test::WWW::Mechanize'                     => '1.14',
        'utf8' => 0,
        'warnings' => 0,
        'WWW::Mechanize'                           => '1.50',
    },
    dist_version_from => "lib/Test/WWW/Mechanize/Mojo.pm",
    license => "perl",
    create_makefile_pl => 'traditional',
    meta_merge =>
    {
        resources =>
        {
            repository => "http://bitbucket.org/shlomif/perl-test-www-mechanize-mojo",
            homepage => "http://web-cpan.berlios.de/modules/Test-WWW-Mechanize-Mojo/",
        },
        keywords =>
        [
        ],
    },

);

$build->create_build_script;
